File: 0859-design-circular-deque/0859-design-circular-deque.java
class MyCircularDeque {
    ArrayDeque<Integer>q;
    int max;
    int curr;
    public MyCircularDeque(int k) {
        q=new ArrayDeque<>();
        max=k;
        curr=0;
    }
    
    public boolean insertFront(int value) {
        if(curr<max){q.offerFirst(value);curr++;return true;}
        return false;
    }
    
    public boolean insertLast(int value) {
        if(curr<max){q.offerLast(value);curr++;return true;}
        return false;
    }
    
    public boolean deleteFront() {
        if(curr>0){q.pollFirst();curr--;return true;}
        return false;
    }
    
    public boolean deleteLast() {
        if(curr>0){q.pollLast();curr--;return true;}
        return false;
    }
    
    public int getFront() {
        if(curr>0){return q.peekFirst();}
        return -1;
    }
    
    public int getRear() {
        if(curr>0){return q.peekLast();}
        return -1;
    }
    
    public boolean isEmpty() {
        return curr==0;
    }
    
    public boolean isFull() {
        return  curr==max;
    }
}

/**
 * Your MyCircularDeque object will be instantiated and called as such:
 * MyCircularDeque obj = new MyCircularDeque(k);
 * boolean param_1 = obj.insertFront(value);
 * boolean param_2 = obj.insertLast(value);
 * boolean param_3 = obj.deleteFront();
 * boolean param_4 = obj.deleteLast();
 * int param_5 = obj.getFront();
 * int param_6 = obj.getRear();
 * boolean param_7 = obj.isEmpty();
 * boolean param_8 = obj.isFull();
 */

File: 0859-design-circular-deque/0859-design-circular-deque.py
from collections import deque

class MyCircularDeque:
    def __init__(self, k: int):
        self.size = k
        self.dq = deque()
        self.curr = 0

    def insertFront(self, value: int) -> bool:
        if self.curr == self.size:
            return False
        self.dq.appendleft(value)
        self.curr += 1
        return True

    def insertLast(self, value: int) -> bool:
        if self.curr == self.size:
            return False
        self.dq.append(value)
        self.curr += 1
        return True

    def deleteFront(self) -> bool:
        if self.curr == 0:
            return False
        self.dq.popleft()
        self.curr -= 1
        return True

    def deleteLast(self) -> bool:
        if self.curr == 0:
            return False
        self.dq.pop() 
        self.curr -= 1
        return True

    def getFront(self) -> int:
        if self.curr == 0:
            return -1
        return self.dq[0]

    def getRear(self) -> int:
        if self.curr == 0:
            return -1
        return self.dq[-1] 

    def isEmpty(self) -> bool:
        return self.curr == 0

    def isFull(self) -> bool:
        return self.curr == self.size


